index.html

```diff
@@ -1028,6 +1061,461 @@ <h2>
+      </h2>
+      <p>
+        A {{GamepadHapticActuator}} corresponds to a configuration of motors or
+        other actuators that can apply a force for the purposes of haptic
+        feedback.
+      </p>
+      <pre class="idl">
+        [Exposed=Window]
+        interface GamepadHapticActuator {
+          readonly attribute FrozenArray&lt;GamepadHapticEffectType&gt; effects;
+          Promise&lt;GamepadHapticsResult&gt; playEffect(
+              GamepadHapticEffectType type,
+              optional GamepadEffectParameters params = {});
+          Promise&lt;GamepadHapticsResult&gt; reset();
+        };
+      </pre>
+      <p>
+        Instances of {{GamepadHapticActuator}} are created with the internal
+        slots described in the following table:
+      </p>
+      <table class="simple">
+        <tr>
+          <th>
+            Internal slot
+          </th>
+          <th>
+            Initial value
+          </th>
+          <th>
+            Description (non-normative)
+          </th>
+        </tr>
+        <tr>
+          <td>
+            <dfn data-dfn-for="GamepadHapticActuator">[[\effects]]</dfn>
+          </td>
+          <td>
+            An empty [=list=].
+          </td>
+          <td>
+            A [=list=] of {{Gamepad/GamepadHapticEffectType}} values
+            representing all the effects supported by the actuator
+          </td>
+        </tr>
+        <tr>
+          <td>
+            <dfn data-dfn-for=
+            "GamepadHapticActuator">[[\playingEffectPromise]]</dfn>
+          </td>
+          <td>
+            `undefined`
+          </td>
+          <td>
+            The {{Promise}} for the playing effect, or `undefined` if no effect
+            is playing.
+          </td>
+        </tr>
+      </table>
+      <dl data-dfn-for="GamepadHapticActuator">
+        <dt>
+          <dfn>effects</dfn> attribute
+        </dt>
+        <dd>
+          <p>
+            Array of {{Gamepad/GamepadHapticEffectType}} values representing all
+            the types of haptic effects that the actuator supports. This
+            property MUST contain all the {{Gamepad/GamepadHapticEffectType}}
+            values that the actuator supports, unless the [=user agent=] does
+            not support playing effects of that type. The same object MUST be
+            returned and MUST NOT change for the lifetime of the
+            {{GamepadHapticActuator}}.
+          </p>
+          <p>
+            The {{GamepadHapticActuator/effects}} getter steps are:
+          </p>
+          <ol>
+            <li>Return [=this=].{{GamepadHapticActuator/[[effects]]}}.
+            </li>
+          </ol>
+        </dd>
+        <dt>
+          <dfn>playEffect()</dfn> method
+        </dt>
+        <dd>
+          <p>
+            The {{GamepadHapticActuator/playEffect()}} method steps are:
+          </p>
+          <ol>
+            <li>If |params:GamepadEffectParameters| does not describe a [=valid
```

Comment:
I think what we actually want here is a switch statement for performing validation of each `type` and a give `params`. Right now, it's you need to go hunting for what do do, which is not great.

I was expecting:
```HTML 
<p>
The {{GamepadHapticActuator/playEffect()}} method steps, called with {{GamepadHapticEffectType}} |type:GamepadHapticEffectType| and {{GamepadEffectParameters}} |params:GamepadEffectParameters | :
</p>
<ol class="algorithm">
  <li>Given the value of type |type:GamepadHapticEffectType|, switch on:
  <dl class="switch">  
    <dt>
      {{GamepadHapticEffectType/"dual-rumble"}}
    </dt>
    <dd>
      <ol>
        <li>let |result be [=validate "dual-rumble" parameters=] passing |params|.
        <li>If result is an an error, return [=a promise rejected=] with error. 
      <ol>
    </dd>
  </dl>
</li>
<!--- other steps --> 
</ol>
```

Reply:
I have these verification steps closer to <dfn>valid effect</dfn>. WDYT?

Reply:
I don't think we should be allowed to even get into the situation where we need a validator (i.e., we don't ever need to run custom validators for each value)... Consider, by merely changing the the WebIDL value for `startDelay` from double to an `unsigned long long` with a default of 0, we've already solved for the validity check without needing any prose (and it already gives you a TypeError for free). 

**I think we can get rid of "valid effect" entirely.**

Reply:
Ok, so the only thing to check is if `weakMagnitude` and `strongMagnitude` are less than or equal to 1 (because it's the only thing WebIDL can't do today).